### PR TITLE
Investment strategies

### PR DESCRIPTION
**Pull request by amy1205**
Created a general interface for Invest contract
Added invest contracts for YVault, Aave and Compound protocols

### CODE REVIEW

**Comment by prateekreddy** [comment]
Handle Ether investments as well

**Review by prateekreddy — CHANGES_REQUESTED** [review]

File: contracts/interfaces/IInvest.sol [hunk 1]
```diff
@@ -2,19 +2,34 @@
...
 
-    function unlockCollateral(
-        bytes32 poolHash,
-        uint256 amount,
-        address from
-    ) external returns (uint256);
+    /**
+     * @dev Used to unlock tokens from available protocol
+     * @param asset the address of underlying token
+     * @param amount the amount of asset
```
[the remark this answers is not in the document]
Comment (amy1205) [reply contracts/interfaces/IInvest.sol:19]: amount of liquidity shares to be unlocked

File: contracts/interfaces/IInvest.sol [hunk 2]
```diff
@@ -2,19 +2,34 @@
...
+    /**
+     * @dev Used to lock tokens in available protocol
+     * @param asset the address of token to invest
+     * @param amount the amount of asset
+     * @return sharesReceived amount of shares received
+     * @return investedTo address of share tokens
+     **/
+    function lockTokens(address asset, uint256 amount)
+        external
+        returns (uint256 sharesReceived, address investedTo);
```
[the remark this answers is not in the document]
Comment (amy1205) [reply contracts/interfaces/IInvest.sol:14]: Yeah right, removed.

File: contracts/invest/InvestAave.sol [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        returns (uint256 amount)
+    {
+        //TODO:
+    }
+
+    function _toAave(address asset, uint256 amount)
+        internal
+        returns (uint256 sharesReceived, address aToken)
+    {
+        require(
```
[the remark this answers is not in the document]
Comment (amy1205) [reply contracts/invest/InvestAave.sol:157]: Right, removed it

File: contracts/invest/InvestCompound.sol [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        address asset,
+        address cToken,
+        uint256 amount
+    ) internal returns (uint256 tokensReceived) {
+        require(cToken != address(0), "Invest: investedTo address");
+        IERC20(cToken).safeTransferFrom(msg.sender, address(this), amount);
+
+        uint256 initialAssetBalance = IERC20(asset).balanceOf(address(this));
+
+        require(ICToken(cToken).redeem(amount) == 0, "Error in unwrapping");
```
[the remark this answers is not in the document]
Comment (amy1205) [reply contracts/invest/InvestCompound.sol:160]: While redeeming, it directly burns the cTokens from msg.sender address

File: contracts/invest/InvestYVault.sol [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        address investedTo,
+        uint256 amount
+    ) internal returns (uint256 tokensReceived) {
+        require(investedTo != address(0), "Invest: investedTo address");
+        IERC20(investedTo).safeTransferFrom(msg.sender, address(this), amount);
+
+        uint256 collateralBefore = IERC20(asset).balanceOf(address(this));
+
+        //withdraw collateral from vault
+        IyVault(investedTo).withdraw(amount);
```
[the remark this answers is not in the document]
Comment (amy1205) [reply contracts/invest/InvestYVault.sol:162]: It burns the tokens from wallet, hence not needed

File: contracts/invest/InvestCompound.sol [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    using SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    address public pool;
+    address public savingsAccount;
+
+    /**
+     * @dev stores the address of contract to invest in
+     */
+    mapping(address => address) investTo;
```
[the remark this answers is not in the document]
Comment (amy1205) [reply contracts/invest/InvestCompound.sol]: Yes, to get the cToken contract address to invest in. I am looking for a registry but to avoid this

**Comment by prateekreddy** [comment]
@amy1205 We can mark all getTokensForShares methods to view as they don't change state.

**Comment by prateekreddy** [comment]
Also probably set referral code stuff to variable may be and admin can set it

**Comment by prateekreddy** [comment]
The tokenshares should instead be stored against msg.sender. The current scheme is attackable because the ownership of the tokens are with user rather than pool contract.

**Pull request merged**